File: prev_exams_and_ans/2012ccc/q5_a_coin_game/ACoinGame.java
import java.util.Scanner;
import java.util.LinkedList;
import java.util.HashSet;
import java.util.Arrays;

public class ACoinGame {
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		String n;
		while (!(n = sc.nextLine()).equals("0")) {
            
            boolean foundFlag = false;
			HashSet<String> visited = new HashSet<>();
			String[] initCurState = sc.nextLine().split(" ");
            
            State initState = new State(); 
            initState.level = 0; 
            initState.curState = initCurState;
			
			// test if the original place is already finished
			if (isFinished(initState)) {
				System.out.println(initState.level);
				foundFlag = true;
				continue ;
			}
			
			visited.add(Arrays.toString(initState.curState));
			
			LinkedList<State> tries = new LinkedList<>();
			LinkedList<State> nextInitialStates = nextVisits(initState);
			tries.addAll(nextInitialStates);
			
			for (State nextState: nextInitialStates) {
				visited.add(Arrays.toString(nextState.curState));
			}
			
			
			while(tries.size() > 0) {
				State nextState = tries.removeFirst();
				if (isFinished(nextState)) {
					System.out.println(nextState.level);
					foundFlag = true;
					break ;
				}
				
				// System.out.println("Finding next visits for " + Arrays.toString(nextState.curState));
				
				LinkedList<State> nextVisits = nextVisits(nextState);
				for (State nextVisit: nextVisits) {
					if (!visited.contains(Arrays.toString(nextVisit.curState))) {
						visited.add(Arrays.toString(nextVisit.curState));
						tries.add(nextVisit);
					}
				}
			}
			
			if (!foundFlag) {
				System.out.println("IMPOSSIBLE");
			}
			
		}
		sc.close();
//		String[] initCurState = {"32", "", "1"};
//		State initState = new State(); initState.level = 0; initState.curState = initCurState; 
//		nextVisits(initState);
//		
//		System.out.println(isFinished(initState));
	}
	
	public static LinkedList<State> nextVisits(State state) {
		LinkedList<State> newStates = new LinkedList<>();
		for (int i = 0; i < state.curState.length; i++) {
			for (int j = i-1; j <= i + 1; j++) {
				if (j < state.curState.length && j >= 0) {
					String slotI = state.curState[i];
					String slotJ = state.curState[j];
					if (!slotJ.equals("")) {
						int jLen = slotJ.length();
						int iLen = slotI.length();
						if ((!slotI.equals("") && slotJ.charAt(jLen - 1) < slotI.charAt(iLen-1)) || (slotI.equals(""))) {
							slotI += slotJ.charAt(jLen - 1);
							slotJ = slotJ.substring(0, jLen - 1);
							String[] newState = Arrays.copyOf(state.curState, state.curState.length);
							newState[i] = slotI;
							newState[j] = slotJ;
							
							// System.out.println(Arrays.toString(newState));
							
							State stateToBeAdded = new State();
							stateToBeAdded.level = state.level + 1;
							stateToBeAdded.curState = newState;
							newStates.add(stateToBeAdded);
						}
					}
				}
			}
		}
		return newStates;
	}
	
	public static boolean isFinished(State state) {
		for (int i = 0; i < state.curState.length; i++) {
			if (state.curState[i].equals("") || Integer.parseInt(state.curState[i]) != i + 1)
				return false;
		}
		return true;
	}
}


class State {
	public int level;
	public String[] curState;
}

File: prev_exams_and_ans/2012ccc/q5_a_coin_game/a_coin_game.py
from collections import deque

def is_finished(state):
    for i in range(0, len(state[1])):
        if state[1][i] == "" or int(state[1][i]) != i + 1:
            return False
    return True

def next(state):
    for i in range(0, len(state[1])):
        for j in range(i-1, i+2, 1):
            if j < len(state[1]) and j >= 0:
                slot_i = state[1][i]
                slot_j = state[1][j]
                if slot_j != "":
                    if (slot_i != "" and slot_j[-1] < slot_i[-1]) or (slot_i == ""):
                        slot_i += slot_j[-1]
                        slot_j = slot_j[:-1]
                        new_pos = state[1].copy()
                        new_pos[i] = slot_i
                        new_pos[j] = slot_j
                        if ','.join(new_pos) not in visited:
                            tries.append([state[0] + 1, new_pos])
                            visited.add(','.join(new_pos))


while True:
    n = int(input())
    if n == 0:
        break

    found = False
    visited = set()
    tries = deque()
    init_pos = [x for x in input().split()]

    init_state = [0, init_pos]

    if is_finished(init_state):
        print(init_state[0])
        found = True
        continue
    
    tries.append(init_state)
    visited.add(','.join(init_state[1]))

    while (len(tries) > 0):
        curr_state = tries.popleft()
        if is_finished(curr_state):
            print(curr_state[0])
            found = True
            break

        next(curr_state)

    if not found:
        print("IMPOSSIBLE")
